Update to CMUCTAT/XBlockStattutor master

Ringenberg opened the pull request.

vvelsen committed:
First commit of what should be the full codebase needed to run the stattutor version of the CTAT XBlock

vvelsen committed:
Much better version of the XBlock using the existing OLI superactivity HTML5 code. We're now executing our tutoring code and mostly the generated HTML fragment is complete

vvelsen committed:
An almost complete migration from our OLI HTML5 stattutor to an XBlock version. All dynamically loadable files should now be loadable on demand and the loading order should be correct now. This version is missing proper jsrender support and the usage of static resources such as images

vvelsen committed:
More code to manage and fix the way an XBlock manages resources such as images, xml and json files. This commit should provide at least a baseline stattutor. It should grade and should have all the correct resources bundled with it

vvelsen committed:
Quick commit to get the latest fixes into the repository that address resources loading issues. This one also fixes a few bugs in ctatloader, which wasn't completely configured to detect the difference between OLI and XBlock

Commit:
Potential fix to the error caused by using jsRender.
It now should read the template from a string stored in a variable and
loaded using add_javascript.

Ringenberg committed:
Fixed a parser error in jsrenderbody.js (unescaped backslash)
Moved jsrenderbody to static/js

Ringenberg committed:
Fixed reference to Instructions.xml from instructions.xml

Ringenberg committed:
Fixed order of loading to get the jsrendered templates to load and execute at
the right time.
Fixed references to point to correct css files for easyui and fixed loading
order.
Merged jsrenderbody.js and fragment1.js.
Renamed fragment2.js to load_resources.js.

Ringenberg committed:
Moved to python3 style print statements. (fully python2 compliant).

Ringenberg committed:
Removed dead public/js/fragment*.js files and added missing css stattutor.css

Ringenberg committed:
Fixed use of window.module (a reserved variable) which solves the CTATFunctionParser not defined error

Ringenberg committed:
Had to hardcode part of the FlashVars[question_file] because not all of the necessary environment variables are set at that point.

Ringenberg committed:
Working StatTutor XBlock.  Grading is supposedly getting reported, but needs testing.

Ringenberg committed:
Deleted stattutor/__init__.pyc

This file should not have been checked in as it is a generated/compiled python file.

Ringenberg committed:
Fixes #1.
Syncronized latest StatTutor.html with body.html and jsrenderbody.js, note
there was more out of date stuff than just the resizable work area.
Use Template and "".format() to do some file reference substitution in
stattutor.py.

Ringenberg committed:
Updated stattutor.js

Ringenberg committed:
Templatized load_resources.js to make filename translation easier.
Moved Instructions.xml up to public and fixed #2.

Ringenberg committed:
Addressing #3: got data set selection and download button mostly working.
It will now start a download for .xls files, but not for other types.  This likely has to do with XBlock understanding only a minimal number of mimetypes.

Ringenberg committed:
Updated to latest ctat.min.js which fixes the problem of text components not going ungraded when modified.

Ringenberg committed:
Updated ctat libraries with fixes to a double loading issue and a fix for setVariable not escaping properly.
Added a fix in load_resources to deal with xblock's ancient jquery.

Ringenberg committed:
Fixed adding templated question panels to the work panel section.
Apparently, the stattutor work area does not exist while the page is loading.

Ringenberg committed:
Changes to get it to load the brd from the xblock.

Ringenberg committed:
Updated brd files to most recent.  Copied from OLI.

Ringenberg committed:
Fixed StatTutor css to override edx css that was causing layout and
usability issues.

vvelsen committed:
Small prep to the python file in preparation to hopefully fix the path resolution issues.

vvelsen committed:
Merge branch 'master' of https://github.com/Mindtoeye/XBlockStattutor

Ringenberg committed:
Update survey.brd to include more hints and buggy feedback messages.

vvelsen committed:
I've instrumented the XBlock to generate a debug file in the /tmp directory so that we can see what it thinks the configured environment is.

vvelsen committed:
Merge branch 'master' of https://github.com/Mindtoeye/XBlockStattutor

vvelsen committed:
Had to change Scope.content to Scope.settings for two of our XBlock parameters otherwise we can't write to them. No idea why.

Ringenberg committed:
New version of ctat.min.js that fixes the radio button highlight problem.

vvelsen committed:
Patches from our template XBlock ported to Stattutor

Ringenberg committed:
Updated brd files fixing some bad hints telling students to go to the wrong panel.

Ringenberg committed:
Merge branch 'master' of github.com:CMUCTAT/XBlockStattutor

Ringenberg committed:
Potential fix for #7 by moving the code that calls jsrender functions to the .ready() handler.

Ringenberg committed:
Using UnderscoreJS for templating instead of jsRender.

Ringenberg committed:
Removed jsrender.min.js

Ringenberg committed:
Trying to remove //localhost from resource urls.

Ringenberg committed:
Added some comments.

vvelsen committed:
Ported the studio code from the template XBlock to the Stattutor version.

vvelsen committed:
Merge branch 'master' of https://github.com/Mindtoeye/XBlockStattutor

vvelsen committed:
Forgot to change all the module instances to ctatmodule to avoid conflicts with Node.

vvelsen committed:
More module confusion fixed.

vvelsen committed:
For some reason we still had a dependency on the src tag. I removed it but we will have to investigate why it was there in the first place.

Ringenberg committed:
Fixed problem_location reference.

Ringenberg committed:
Added .gitignore based on https://github.com/github/gitignore/blob/master/Python.gitignore so that junk is ignored.

Ringenberg committed:
Ported grading from XBlockCTAT.

Ringenberg committed:
Added missing import.

Ringenberg committed:
Save and restore should be enabled now.

Commit:
Updated to latest ctat.min.js

Pulled the latest from the HEAD branch of HTML5 and rebuilt the js file
to be included here.

Commit:
Fixed a typo.

Fixed a small typo from my slightly dirty version of HTML5.

Ringenberg committed:
Added save-and-restore to ctatloader.js

Ringenberg committed:
Trying to get studio editing working so that the module can be changed.

Ringenberg committed:
Got editing module parameters working.

Ringenberg committed:
Fixed bad reference in ctatxblock.html.  It should set ctatmodule not stattutor_module.
This should fix #8.

Ringenberg committed:
Merge branch 'master' into grading

Ringenberg committed:
Started work on getting the stattutor inside an iframe.
This adds a bunch of files that are required for the standalone/iframe version of the tutor.

Ringenberg committed:
Working on getting bi-directional communication to work between edx and ctat in an iframe.

Ringenberg committed:
Update to get saveandrestore to work.

Ringenberg committed:
First stage cleanup.  Removing no longer needed files.

Ringenberg committed:
Fixing up studio editing.

Ringenberg committed:
Fixed references in studio edit page.

Ringenberg committed:
Fixing more reference problems in studio edit.

Ringenberg committed:
Added *~ to .gitignore so Emacs backups are ignored.

Ringenberg committed:
Advanced version number.

Ringenberg committed:
Now have a simple combo box for selecting the module.

Ringenberg committed:
Fixed a typo and added some necessary path information.

Ringenberg committed:
Fixed bad python code.

Ringenberg committed:
Fixed relative path information.

Ringenberg committed:
code cleanup

Ringenberg committed:
Merge branch 'master' into grading

Ringenberg committed:
Removed bad confict marker.

Ringenberg committed:
Added weight, which seem to be required, not optional.

Ringenberg committed:
Merging in grading and fixing ctat_grade return.

Ringenberg committed:
Increased version number to indicate that grading seems to work.

Ringenberg committed:
Cleaning up of code and small changes to more closely match XBlockCTAT.

Commit:
Added .gitattributes files to help with line endings

Ringenberg committed:
Merge branch 'master' of github.com:CMUCTAT/XBlockStattutor

Ringenberg committed:
Removed extranious end tag

Ringenberg committed:
Added missing CTATXConfig.js file.

Ringenberg committed:
Updated to use CTATLMS based on changes from XBlockCTAT.

Ringenberg committed:
Updated version number.

Ringenberg committed:
Changes based on Dawn's comments.
Conduct Analysis no longer has a check box.
Added additional instruction text.

Ringenberg committed:
Fixed Our Answer size problem.
Removed some dead code.
Updated to latest CTAT code.

Ringenberg committed:
Increased version number.

Ringenberg committed:
Updated to latest pre-release version of ctat.

Ringenberg committed:
Minor updates to bring in line with OLI version, mostly fixing css

Ringenberg committed:
Discount the number of possible steps by 1 to try to fix grading #11

Ringenberg committed:
Another stap at #11, this time using the max number of correct.

Ringenberg committed:
Work around 1 for #11.

Ringenberg committed:
Commented out some errors

Ringenberg committed:
Added some error handling and made sure both numbers for scoring are floats and added check to try to reduce double entries. Fixes for #11

Ringenberg committed:
updated version number

Ringenberg committed:
Create README for problem_files directory. (#12)

Ringenberg committed:
Pylint cleanup (#13)

* Pylint now reporting a score of 10/10

* Python code now passes pep8 and pylint.

* Updated to the official CTAT 4.0.0 release. Updated documentation and version number.

Ringenberg committed:
Added license

Ringenberg committed:
Added display_name instance variable. (#18)

Ringenberg committed:
Tepletized width and height. (#19)

Ringenberg committed:
Updated version number.

To reflect fixing some bugs, the version number has be incremented.

stvstnfrd committed:
Re-enable pylint import errors

Yeah, it makes sense to turn it off for XBlock since we know that
dependency is fullfilled elsewhere, but we can turn it right back on
afterward, just in case we add more imports later.

stvstnfrd committed:
Remove commented-out code

Static assets like JS and HTML are still sent across the wire to clients,
even if the code is commented out.

Ringenberg committed:
Removed some dead files.

Ringenberg committed:
Added a link for XBlock.

Ringenberg committed:
Update .gitattributes

Update .gitattributes to recognize our .brd files as XML

Ringenberg committed:
Brought content in line with current OLI content.
Made some adjustments to make it easier to share resources between the two
environments.

Ringenberg committed:
Fixed the name of one of the logging parameters.

Ringenberg committed:
Version number bump.

Ringenberg committed:
Merge branch 'master' of github.com:CMUCTAT/XBlockStattutor

Ringenberg committed:
Updated problem import directions.

Ringenberg committed:
Merge pull request #22 from stvstnfrd/pylint

Re-enable pylint import errors

Ringenberg committed:
Merge pull request #24 from stvstnfrd/remove-comments

Remove commented-out code

Ringenberg committed:
Potential fix for Issue #11 by including the contents of the brd file instead of requesting it.

Ringenberg committed:
Merge pull request #26 from CMUCTAT/Issue14

Potential fix for Issue #11 by including the contents of the brd file

Ringenberg committed:
Documentation and added problem white list.

Ringenberg committed:
Reverting ctat.min.js

Ringenberg committed:
Fixed some data file references.

Ringenberg committed:
Simplified logging parameters
Reduced number to be maintained and get more information directly from the environment.

Ringenberg committed:
Put some sanity checks on 'Logging' so that Studio previews do not log.

Ringenberg committed:
Merge pull request #27 from CMUCTAT/Security_and_Documenting

Security and documenting

Ringenberg committed:
Added instruction on how to register mime types. 

Added instruction in Readme.md that explain how to register the file types used in StatTutor supported statistical packages.
This should cover Issue #6

Ringenberg committed:
Added instruction on adding new content

This adds some basic instruction on importing new modules from OLI. This should cover #9.

Ringenberg committed:
Adding some logging configuration and extra grading safety

Ringenberg committed:
Fixes from testing.

Ringenberg committed:
Removed default log server entry.

jsewall committed:
Port Norm's 2016-09-12 changes to OLI to XBlock.
His note: it looks like I had a mistake in the brd for m2_body_image_2q  that was preventing it from working correctly in OLI (and I suspect that is what was breaking it in edx.  I have corrected the .brd (and tweaked the .xml) in the version on the experimental/branches/ctat repo and it.s now working correctly in OLI.  I attempted to load and test on ctat-new, but none of the stattutors are working for me (getting an error in finding the brd).
As I was fixing this, I also discovered that the m5_extra_credit_2q . the path to the json file for data was off, so it wasn.t showing the data.  I also corrected that in the experimental branch.

jsewall committed:
Change selection names on buggy links from q3 to q2. Fix unquoted string argument.

jsewall committed:
1. Fix formula parser to better handle string literals.2. Revise success msg, fix syntax errors in buggy message generator function calls.

Ringenberg committed:
Fixed scroll bars in hint window not being interactive.

jsewall committed:
Merge remote-tracking branch 'origin/FixScrollBars' into Revise2QuestionProblems2and5 to get css fix for scroll bars.

jsewall committed:
Revise success msg in final reflect-on-conclusion step to guide student to summarize step.

Ringenberg committed:
Merge pull request #29 from CMUCTAT/Revise2QuestionProblems2and5

Feedback in 2-question problems; guidance to Summarize; formula evaluator for string literals

Mindtoeye committed:
Added the ability to use the built-in logging capabilities to send our OLI/Datashop logs to. When selected logging activity will end up in: /edx/var/log/tracking/tracking.log

Ringenberg committed:
Added x2js library.

Ringenberg committed:
Added some safety in studio for SDK environment.

Ringenberg committed:
Fixed brd loading and linked to correct xml2json library.

Ringenberg committed:
Now using compiled in version of xml2json.

Ringenberg committed:
Merge pull request #30 from CMUCTAT/NativeEdXLogging

Native EdX logging

Ringenberg committed:
Update version number

Ringenberg committed:
Unified logging with XBlockCTAT to log xml.

Ringenberg committed:
Fix logging error in studio.

Ringenberg committed:
Added some documentation.

Ringenberg committed:
Updated version number and ctat.min.js

Ringenberg committed:
Merge branch 'NativeEdXLogging'

Ringenberg committed:
Merge branch 'master' of github.com:CMUCTAT/XBlockStattutor

Ringenberg committed:
Removed need for edx:/localhost fake url.  Blank will now suffice for log_service_url.

Mindtoeye committed:
Incorporated Jonathan's hint fix.

Ringenberg committed:
Addressing #32. Converted greater-than-or-equal to &ge;

Ringenberg committed:
Converted ... to &hellip;

Ringenberg committed:
Double escape html characters.

Ringenberg committed:
Merge pull request #33 from CMUCTAT/FixGEIssue32

Fix for #32

Mindtoeye committed:
Changed the code to use xml2json.min.js locally in the Stattutor version of CTATLMS.js. Beforehand the code stored a JSON.stringified version of the incoming log XML.

Mindtoeye committed:
Patched the json generation code to include a key per message that is made from the problem name and current step/question.

Mindtoeye committed:
Merge remote-tracking branch 'remotes/origin/NativeEdXLogging'

Commit:
Enhance ctat.min.js to provide the last success message in Variable Table entry success_message; revise the .brd files to use the new feature by including <%=success_message%> in success messages on navigation steps.

Commit:
Supply missing m4_cell_phone/ element in paths to images. Change UTF-8 chars to HTML entities. Revise navigation messages for Q3 and Summarize.

Commit:
With AuthoringTools/trunk/HTML5 revs 24260 24285 24295 24321 24332 24354 24364 on branch XBlockStattutor: revised go-to-state using browser storage; use selection to label events logged to edX. Changes merged from master to static/ for event logging mods.

Commit:
Add missing file, copied from ../XBlockStattutor-olicdn/.

jsewall committed:
Merge pull request #34 from CMUCTAT/SuccessMsgAndDSSessionLog

Preserve the last success message so navigation rubrics can be appended instead of overwriting.

Commit:
1. In ctat.min.js, avoid saving or logging during problem restore.
2. In xml2json, remove code that tries to put the constructor in define() or exports(): always set the global X2JS needed by Initialize_CTATXBlock.
3. In stattutor.py, permit blank log_service_url, so can log only to edX.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        return [
+            ("StattutorXBlock",
+             """<vertical_demo>
+                <stattutor width="900" height="750"/>
```
This is outdated, right?
Height/width are no longer configurable on an xblock-by-xblock basis.

Providing useful scenarios makes workbench development easier by demoing varying use cases.
What could be configured here, `problem`, `weight`, etc.?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        Return the stored problem state to reconstruct a student's progress.
+
+        Args:
+          self: the StatTutor XBlock.
```
Don't document the `self` argument.
This is standard Python syntax.

stvstnfrd replied:
This comment applies to all other occurrences too.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+          A JSON object with a 'result' field with a payload indicating the
+          success status.
+        """
+        if data.get('state') is not None:
```
The Pythonic way to check for existence of a key in a collection is:

```python
if 'state' in data:
    # whatever
```

Ringenberg replied:
data is a JSON object not a python collection. When I tested it the python way, it gave me errors.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+
+    @XBlock.json_handler
+    def ctat_save_problem_state(self, data, dummy_suffix=''):
+        """Called from CTATLMS.saveProblemState.
```
Begin the comment text on a newline:

```python
"""
Called from CTATLMS.saveProblemState
"""
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+
+    def max_score(self):
+        """ The maximum raw score of the problem. """
+        # For some unknown reason, errors are thrown if the return value is
```
What kind of errors?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+
+    # **** Grading variables ****
+    # All of the variable in this section are required to get grading to work
+    # according to EdX's documentation.
```
Which documentation?

stvstnfrd commented on stattutor/public/js/CTATXConfig.js:
```diff
@@ -0,0 +1,12 @@
+/** set the environment */
+var CTATTarget="XBlock"; // this should eventually be removed
+var CTATXConfig = window.parent.CTATConfig; // grab from encompasing environment
+setVariable = window.parent.CTATXBlock.post.set_variable;
```
Missing `var` keyword?

stvstnfrd commented on stattutor/public/js/CTATXConfig.js:
```diff
@@ -0,0 +1,12 @@
+ */
+
+/** set the environment */
+var CTATTarget="XBlock"; // this should eventually be removed
```
When/why?

stvstnfrd commented on stattutor/public/html/StatTutor.html:
```diff
@@ -0,0 +1,634 @@
+<script src="../js/stattutor.js"></script>
+
+<script>
+//useDebuggingBasic=true;
```
Remove commented code.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        logserver = bleach.clean(data.get('logserver').strip(), strip=True)
+        logging = bleach.clean(data.get('logging'), strip=True)
+        if statmodule in self.problems.keys():
+            self.problem = statmodule
```
Which is the more apt name, `problem` or `statmodule`?
We should chose one name and stick with it, instead of changing it from front-end to back-end.
This makes searching and debugging easier.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            self.problem = statmodule
+        else:
+            status = 'failure'
+            messages.append("invalid module")
```
Why is this an array if it will never have more than one item?
Shouldn't we just say

```python
result['message'] = 'invalid module'
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            status = 'failure'
+            messages.append("invalid module")
+        if logging.lower() == "true":
+            self.logtype = True
```
This name doesn't match the type.
Should this be something like `enable_logging`?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            messages.append("invalid module")
+        if logging.lower() == "true":
+            self.logtype = True
+            if len(logserver) > 0:
```
This means you can never delete/remove the `log_url`, once set,
since it will ignore blank entries.
You probably don't want this if check.
Actually, why are we only saving the URL if logging is enabled.
As a user, I would be surprised if I made changes to the URL and they weren't saved, even if I was temporarily disabling it (what if I want to disable logging and set the URL to be empty?).

stvstnfrd replied:
This whole block could be cleaned up as:

```python
self.log_url = logserver
self.enable_logging = logging == 'true'
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                self.log_url = logserver
+        else:
+            self.logtype = False
+        ret = {'result': status}
```
Please use a more descriptive variable name than `ret`; this is more difficult to debug.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        Returns:
+          A JSON object with a 'result' and a 'state' field.
+        """
+        return {'result': 'success', 'state': self.saveandrestore}
```
Break up dictionaries into multiple lines:

```python
return {
    'result': 'success',
    'state': self.save_and_restore,
}
```

This is easier to read, debug, and maintain.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+          success status.
+        """
+        if data.get('state') is not None:
+            self.saveandrestore = bleach.clean(data.get('state'))
```
Since you're explicitly checking for the `state` key (looking before you leap),
there's no need to call the safe `dict.get` method.
Instead, we can index directly into the collection.

```python
state = data['state']
state = bleach.clean(state)
self.save_and_restore = state
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        if data.get('state') is not None:
+            self.saveandrestore = bleach.clean(data.get('state'))
+            return {'result': 'success'}
+        return {'result': 'failure'}
```
We prefer that even single item collections be broken up into multiple lines;
it makes it easier/cleaner to add additional items later.

```python
return {
    'result': 'failure',
}
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+
+        Args:
+          self: the StatTutor XBlock.
+          data: A JSON object with a 'state' field with a payload of the blob
```
Curiously, what kind of data is stored in this blob?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            if len(logserver) > 0:
+                self.log_url = logserver
+        else:
+            self.logtype = False
```
By convention, Python code should always be `snake_cased`.

This applies to occurrences throughout this file.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        status = 'success'
+        messages = []
+        statmodule = bleach.clean(data.get('statmodule'), strip=True)
+        logserver = bleach.clean(data.get('logserver').strip(), strip=True)
```
Similarly, can we chose a single name and use it consistently?
Either `log_server` or `log_url`...

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            logging='checked' if self.logtype else '',
+            logserver=self.log_url))
+        studio_js = self.resource_string("static/js/ctatstudio.js")
+        frag.add_javascript(unicode(studio_js))
```
This should instead use `add_javascript_url`. Calling `resource_string` reads the entire file into memory.
We should avoid doing this unless necessary.

http://edx.readthedocs.io/projects/xblock/en/latest/fragment.html#xblock.fragment.Fragment.add_javascript_url

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        # only change score if it increases.
+        # this is done because corrects should only ever increase and
+        # it deals with issues EdX has with grading, in particular
+        # the multiple identical database entries issue.
```
Can you provide some context on these issues?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            # pylint: disable=broad-except
+            # The errors that should be checked are django errors, but there
+            # type is not known at this point. This exception is designed
+            # partially to learn what the possible errors are.
```
Now that you've been running this code,
what errors are possible here?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                    corrects = 0  # check for invalid value
+            if data.get('max_value') is not None:
+                max_val = int(data.get('max_value'))
+                if not math.isnan(max_val) and max_val > 0:
```
I don't think this `isnan` check does what you expect...

In Python, `NaN` is a float (`nan = float('nan')`).
Since you're previously casting `max_value` to an `int`,
one of two things would happen here:
1. The number is an `int`, the cast succeeds, so `math.is_nan` won't be false.
2. The number is some form of `NaN`, the cast throws an exception (`int('foo')`), so we never reach the `NaN` check anyway.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                if math.isnan(corrects):
+                    corrects = 0  # check for invalid value
+            if data.get('max_value') is not None:
+                max_val = int(data.get('max_value'))
```
Why abbreviate to `max_val` here?
This is easier to read and maintain if we preserve the same name throughout.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                    # only update if a valid number
+                    self.max_problem_steps = max_val
+        except ValueError as int_err:
+            return {'result': 'fail', 'error': 'Bad grading values:' +
```
I suspect it would be helpful to use a more descriptive error message, like `max_value must be a number/integer`. As is, the message is `ValueError: invalid literal for int() with base 10: 'foo'`.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            return {'result': 'fail', 'error': 'Bad grading values:' +
+                    unicode(int_err)}
+        # only change score if it increases.
+        # this is done because corrects should only ever increase and
```
"Should"? What scenario could cause it to decrease?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                corrects = int(data.get('value'))
+                if math.isnan(corrects):
+                    corrects = 0  # check for invalid value
+            if data.get('max_value') is not None:
```
`if 'max_value' in data:`

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        """
+        self.attempted = True
+        corrects = 0
+        try:
```
Try blocks should be as limited as possible and not wrap any more lines than necessary:

```python
if 'value' in data:
    try:
        corrects = int(data['value'])
    except ValueError as int_err:
        return whatever
    if 'max_value' in data:
        # etc...
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        try:
+            if data.get('value') is not None:
+                corrects = int(data.get('value'))
+                if math.isnan(corrects):
```
See note below about `math.isnan`..

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            data['user_id'] = self.runtime.user_id
+            data['component_id'] = unicode(self.scope_ids.usage_id)
+            self.runtime.publish(self, "ctatlog", data)
+        # General mechanism to catch a very broad category of errors.
```
Which errors specifically?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    @XBlock.json_handler
+    def ctat_log(self, data, dummy_suffix=''):
+        """Publish log messages from a CTAT tutor to EdX log."""
+        if data.get('event_type') is None or\
```
Don't use slashes for line continuation.

```python
if 'event_type in data and 'action' in data and 'message' in data:
    return {
        # whatever
    }
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        # Add javascript initialization code
+        frag.add_javascript(self.resource_string("static/js/Initialize_CTATXBlock.js"))
+        # Add the xml2json library here because someone has a problem if it lives somewhere else instead
+        frag.add_javascript(self.resource_string("static/js/xml2json.min.js"))		
```
There are a few important, performance releated reasons to use `add_javascript_url` over `add_javascript`
1. Uses less memory. Calling `resource_string` forces the entire file to be read into memory on each load, whereas adding via URL allows the file to be served directly by the server.
2. Delegating to the webserver/nginx will be faster.
3. Improved caching. As an external file, browsers/servers can cache this file. By adding it inline to the HTML document, the contents are never cached and are instead sent across the wire on every page load.

There are instances where it makes sense to add content directly inline (usually when we're modifying the file in flight), but when we're passing the file through untouched (like here), we should add it as an external file.

http://edx.readthedocs.io/projects/xblock/en/latest/fragment.html#xblock.fragment.Fragment.add_javascript_url

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        ))
+        # Add javascript initialization code
+        frag.add_javascript(self.resource_string("static/js/Initialize_CTATXBlock.js"))
+        # Add the xml2json library here because someone has a problem if it lives somewhere else instead
```
Can you clarify? Who is "someone" and where else would it live?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            problem_description=self.get_local_resource_url(description)
+        ))
+        # Add javascript initialization code
+        frag.add_javascript(self.resource_string("static/js/Initialize_CTATXBlock.js"))
```
`add_javascript_url`

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            logtype=self.logtype,
+            log_url=self.log_url,
+            question_file="data:file/brd;base64," +
+            base64.b64encode(self.resource_string(brd)),
```
Per previous discussion, we agreed to revert this to serving the BRD file directly (with custom mimetype support added to the platform), since we don't want to read these entire files into memory on each page load.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            if not sdk_usage else usage_id,
+            problem_name=self.problem,
+            block_type=unicode(usage_id.block_type)
+            if not sdk_usage else usage_id,
```
Inline ternary statements are confusing enough in Python, this is worse being split onto multiple lines.
This is easier to maintain by explicitly assigning values:

```
if sdk_usage:
    block_type = unicode(usage_id.block_type)
    # etc...
else:
    block_type = usage_id
    # etc...
config.format(
    block_type=block_type,
    # etc...
)
```

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+            if hasattr(self.runtime, 'anonymous_student_id')
+            else 'bogus-sdk-id',
+            # class
+            course=unicode(usage_id.course) if not sdk_usage else usage_id,
```
Why would we use `usage_id` as the default for org, course, run, etc?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        sdk_usage = isinstance(usage_id, basestring)
+        frag.add_javascript(config.format(
+            # meta
+            guid=str(uuid.uuid4()),
```
Here you use `str`, elsewhere you use `unicode`;
why the discrepancy?

stvstnfrd replied:
Why do we keep generating new guids? What is this used for?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        frag = Fragment(html.format(
+            # Until the iframe srcdoc attribute is universally supported
+            # a valid xblock generated url has to be passed into
+            # ctatxblock.html.  Internet Explorer does not support srcdoc.
```
Which IE versions, all or just some?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+        # from a remote machine fail completely.
+        return self.strip_local(self.runtime.local_resource_url(self, url))
+
+    # **** XBlock methods ****
```
Let's drop this comment; it adds nothing to the conversation.
We know these are XBlock methods; they're methods defined in the XBlock's class definition...

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    # This section includes variables necessary for storing partial
+    # student answers so that they can come back and work on a problem
+    # without worrying about loosing progress.
+    saveandrestore = String(help="Internal data blob used by the tracer",
```
What kind of data is stored here? Why the opaque blob?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                     scope=Scope.settings)
+
+    # None, ClientToService, ClientToLogServer or OLI
+    # Set to "ClientToService" to activate logging.
```
This describes strings, but the data type is s `bool`...

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    log_url = String(help="URL of the logging service, used to indicate " +
+                     "where the server is that will receive the log messages",
+                     default="",
+                     scope=Scope.settings)
```
- `Scope.settings` means that the value is unique for _every_ single
    instantiation of the XBlock. Right now, the default URL is empty.
    If/when you decide to update this to a real logging server URL, you
    would need to change it, manually, for every instance of the
    StatTutor XBlock in each course. This most likely not your intent.
    You probably want to change it once and have that value reflected
    everywhere.
  - While there is a scope called `Scope.content` that would apply the
    setting to any/all copies of the XBlock, this is problematic because
    the logging URL is instructor-configurable; a mistake by one
    instructor would affect everyone.
  - My reccommendation is to hard-code this value directly in the code
    (defining a constant, of course).

`log_url = 'https://foo.bar'`

http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/concepts/fields.html#predefined-scopes

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    # Generate and store a dictionary of the available problems.
+    # (AKA the problem whitelist)
+    problems = {}
+    for pf_dir in pkg_resources.resource_listdir(__name__,
```
We should move this whole block into its own helper method, instead of inlining it directly in the class definition.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    # not exist.
+
+    # **** Basic interface variables ****
+    # All of the variable in this section are required to get the tutors to run
```
I'm not really sure what this comment is trying to say.
I would hope all variables are needed. If I had previously suspected that a variable was not needed, a specific explanation is helpful. A blanket statement like this doesn't improve readability.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+              "option point values."),
+        values={"min": 0, "step": .1},
+        scope=Scope.settings
+    )  # weight needs to be set to something, errors will be thrown if it does
```
What kinds of errors? Where?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    icon_class = String(default="problem", scope=Scope.content)
+    score = Integer(help="Current count of correctly completed student steps",
+                    scope=Scope.user_state, default=0)
+    max_problem_steps = Integer(
```
What determines the number of problem steps? Does it vary problem to problem?
Could you have multiple instances of problem 1 with varying numbers of steps?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+    # loaded images and some of the interactive elements without causing
+    # side scrolling scrollbars to appear.  They are set here instead of
+    # hard coding them into ctatxblock.html to make it easier for EdX
+    # administrators to modify them if they wish without having to scour
```
That was actually my point from my original request [1]...
If it only looks right with one set of dimensions,
we _don't_ want to change it, ever.
So let's just hard-code it directly in the HTML template.

I also suggested that it may fit in 800px, which as you point out, has edge-case issues,
so I have no objection to keeping it at 900px.

[1] https://github.com/CMUCTAT/XBlockStattutor/pull/21

stvstnfrd commented on stattutor/static/js/xml2json.min.js:
```diff
@@ -0,0 +1 @@
+(function(a,b){if(typeof define==="function"&&define.amd){define([],b);}else{if(typeof exports==="object"){module.exports=b();}else{a.X2JS=b();}}}(this,function(){return function(z){var t="1.2.0";z=z||{};i();u();function i(){if(z.escapeMode===undefined){z.escapeMode=true;}z.attributePrefix=z.attributePrefix||"_";z.arrayAccessForm=z.arrayAccessForm||"none";z.emptyNodeForm=z.emptyNodeForm||"text";i ...
```
What is the license of this library?

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+implemented for OLI (http://oli.cmu.edu/).
+"""
+
+import re
```
Please alphabetize imports.

stvstnfrd commented on stattutor/static/js/ctatstudio.js:
```diff
@@ -0,0 +1,36 @@
+    });
+    // Add "Cancel" button click event listener.
+    $(element).find('.cancel-button').bind('click', function() {
+	if (runtime.notify)
```
Always wrap Javascript blocks in curly braces, even when just a single line.
This is an all-too-often class of errors, where a block is modified, an extra line is added, but it mistakenly falls outside the `if` block.

```javascript
if (false)
    alert('This will never print');
    alert("This shouldn't print, but it will");
```

stvstnfrd commented on stattutor/static/js/ctatstudio.js:
```diff
@@ -0,0 +1,36 @@
+		runtime.notify('save', {state: 'end'});
+	});
+    });
+    // Add "Cancel" button click event listener.
```
Comments should describe why/how code does something, not what it does.
Most of the inline comments in this function are just repeating exactly what the code does.

It should be clear that `(.cancel-button).bind('click',` binds a click event to the cancel button;
the code is already descriptive enough.
Not only does this add noise to the signal, but by tying the the comments to the underlying implementation, it's one more thing to maintain..

stvstnfrd commented on stattutor/static/js/Initialize_CTATXBlock.js:
```diff
@@ -0,0 +1,113 @@
+ * @param runtime - provided by EdX
+ * @param element - provided by EdX
+ */
+function Initialize_CTATXBlock(runtime,element) 
```
It looks like you massively reformatted this file since I saw it last.
Why does it follow a different coding convention than the rest of your JS?

https://google.github.io/styleguide/jsguide.html
http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/javascript_guidelines.html

stvstnfrd commented on stattutor/static/js/Initialize_CTATXBlock.js:
```diff
@@ -0,0 +1,113 @@
+				type: "POST",
+				url: runtime.handlerUrl(element, 'ctat_grade'),
+				data: JSON.stringify({'value': correct_step_count,
+									  'max_value': total_step_count}),
```
It looks like you're mixing tabs and spaces? 
This causes the code to display differently in different contexts.

All edX projects use soft tabs (spaces) everywhere; in JavaScript soft tabs should be 4 characters in width.

stvstnfrd commented on stattutor/static/js/Initialize_CTATXBlock.js:
```diff
@@ -0,0 +1,113 @@
+	
+    $('.stattutor').on("load", function() 
+	{
+		this.contentWindow.CTATTarget = "XBlock"; // needed for ctatloader.js
```
Maybe it's the hard tabs, but this looks doubly-indented?

stvstnfrd commented on stattutor/static/js/CTATConfig.js:
```diff
@@ -0,0 +1,40 @@
+/**
+ * We prefabricate a set of flashvars such that the
```
"flashvars"?

stvstnfrd commented on stattutor/static/html/ctatxblock.html:
```diff
@@ -0,0 +1 @@
+<iframe id="ctatxblockframe" class="stattutor" src="{tutor_html}" width="{width}" height="{height}"></iframe>
```
By specifying an ID, you can't place more than one XBlock on the page at a time.

stvstnfrd commented on stattutor/public/js/jsrender.min.js:
```diff
@@ -0,0 +1,4 @@
+/*! JsRender v1.0.0-rc.69 (Beta - Release Candidate): http://jsviews.com/#jsrender */
```
What's the license for this file?

stvstnfrd commented on stattutor/public/js/ctat.min.js:
```diff
@@ -0,0 +1,1126 @@
+
+ Copyright (c) Jon Nylander
+*/
+var $jscomp={scope:{}};$jscomp.defineProperty="function"==typeof Object.defineProperties?Object.defineProperty:function(a,b,e){if(e.get||e.set)throw new TypeError("ES3 does not support getters and setters.");a!=Array.prototype&&a!=Object.prototype&&(a[b]=e.value)};$jscomp.getGlobal=function(a){return"undefined"!=typeof window&&window===a?a:"undefined"!=typeof global&&null!=global?global:a};$jscom ...
```
Can you include a non-minified copy of this file in the repo?
It's impossible to review/edit as-is.

stvstnfrd commented on stattutor/public/html/StatTutor.html:
```diff
@@ -0,0 +1,634 @@
+<script src="../js/ctat.min.js"></script>
+<script src="../js/stattutor.js"></script>
+
+<script>
```
Can't this be moved to an external file too?

stvstnfrd commented on stattutor/public/html/StatTutor.html:
```diff
@@ -0,0 +1,634 @@
+
+<script src="https://code.jquery.com/jquery-3.0.0.min.js"></script>
+<script src="https://cdnjs.cloudflare.com/ajax/libs/jsrender/0.9.73/jsrender.min.js"></script>
+<script>
```
Move to external file?

stvstnfrd commented on stattutor/public/html/StatTutor.html:
```diff
@@ -0,0 +1,634 @@
+	<link rel="stylesheet" href="../css/CTAT.min.css" />
+	<link rel="stylesheet" href="../css/stattutor.css" />
+
+<style type="text/css">
```
Forgot to delete this?

stvstnfrd commented on stattutor/public/html/StatTutor.html:
```diff
@@ -0,0 +1,634 @@
+    <head>
```
Why does this file start with `<HEAD>` and not `<HTML>`?

stvstnfrd commented on stattutor/public/css/CTAT.min.css:
```diff
@@ -0,0 +1 @@
+.CTATJumble--item,.unselectable{user-select:none;-khtml-user-select:none}.CTATHintWindow,.CTATJumble--item,.unselectable{-khtml-user-select:none}.CTAT-done-button:focus,.CTAT-gen-component:focus,.CTAT-hint-button:focus,.CTATHintWindow--button{outline:0}.CTAT-svg,.CTATHintWindow,.CTATSkillWindow{box-sizing:border-box}input.CTAT--highlight[type=text],input.CTAT--hint[type=text],textarea.CTAT--highl ...
```
Please include a non-minified copy of this file.

stvstnfrd commented on stattutor/stattutor.py:
```diff
@@ -0,0 +1,386 @@
+                     default="m1_survey", scope=Scope.settings)
+
+    # **** CTATConfiguration variables ****
+    # These should be the only variables needed to set up logging.
```
?

Commit:
Put copies of data files in directories for shortened modules, so they'll be found for them, too.

Commit:
Fix for bugs found in CTATExampleTracerTracer: Success messages were getting interpolated twice. If they included a formula using the new success_message variable, then the variable's value would appear a second time.

Commit:
Remove obsolete module m2_X_body_image/.

jsewall committed:
Merge pull request #35 from CMUCTAT/FixesForLagunitaDec2016

Fixes for lagunita dec2016

jsewall committed:
Fix some misspellings.

Commit:
Load .brd file through ordinary Ajax GET. N.B.: you must add .brd as 'text/xml' to the types listed in the add_mimetypes() functions defined in /edx/app/edxapp/edx-platform/[lc]ms/startup.py (or at least it works if you do).

Commit:
1. Removed unused CTATXConfig.js. 2. No longer put saved state in CTATConfig. Instead, just test its length for problem_state_status.

jsewall committed:
Merge pull request #36 from CMUCTAT/OrdinaryRetrievalForBRD

Ordinary retrieval for brd

jsewall committed:
Add .brd to mime-types in Configuring download types.

Revised instructions in Configuring download types to include CTAT's .brd (behavior recorder diagram) files. Because the client now retrieves these via ordinary Ajax calls, the change to startup.py in /edx/app/edxapp/edx-platform/cms and .../lms is now required.

Commit:
Add problem check event.

jsewall committed:
Merge pull request #38 from Stanford-Online/dcadams/add_problem_check_event2

Add problem check event.

darrensiegel committed:
Addressed code review comments

darrensiegel committed:
corrected logic in data missing check

darrensiegel committed:
removed staticmethod annotation and moved method declaration

jsewall committed:
Merge pull request #39 from darrensiegel/master

Addressed code review comments

Mindtoeye committed:
Changed our code to remove any url from the studio page. At this point you can turn logging only on and off from studio. The log url is empty by default in Python. If a researcher wants they can still force a url by changing the code.

Mindtoeye committed:
Fixed an image resolution issue. This is not an optimal fix but since a study that depends on it is going live soon it's the best we can do. Also patched the logging code to provide more log events and will now included problem_check events for steps that are incorrect.

Commit:
Add feedback provided by Stanford. To maintain, use massProductionTable.xlsx as table, individual _template.brd files.

Commit:
Add feedback provided by Stanford. String formatting functions in stattutor.js.

Commit:
Add feedback from Stanford to Concepts StatTutors.

Commit:
Revise problem_files/ to put all mass production files into new ./massProduction/ because XBlockStattutor cannot handle more than one brd in each subdir.

jsewall committed:
Merge pull request #40 from CMUCTAT/AddFeedback

Add feedback recommended by Stanford.
